Remove the appendix object when JIT calls linkToStaticSpecial

babsingh opened the pull request:
JIT will call `linkToStatic` for unresolved `invokedynamic` and `invokehandle`
sequences. For `invokedynamic` and `invokehandle`, the `appendix` object should not
be appended to the stack if it is null. For the unresolved cases, the JIT does
not know the value of the `appendix` object prior to calling `linkToStatic`. For
the JIT paths, the `appendix` object should be removed from the stack if the
`_sendMethod` does not expect it. The `MemberName` object is located at the top
of the stack. The `appendix` object is located before the `MemberName` object.

Shifting by 2 during the `memmove` will remove the appendix object.

Depends: #12549

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh commented:
`J9ROMMethod->argCount`: number of arguments.
`_currentThread->tempSlot`: number of stack slots needed for the arguments.

So, `_currentThread->tempSlot != J9ROMMethod->argCount`.

@gacholio Is there an easy way to derive stack slots for a `J9ROMMethod`? Or does the JIT need to provide `argCount` in another temp slot?

fyi @liqunl @jdmpapin @nbhuiyan

liqunl commented:
The JIT get the number of parm slots (excluding receiver's) from method signature using `jitParseSignature`

babsingh commented:
`jitParseSignature` evaluates both `paramElements` (== `J9ROMMethod->argCount`) and `paramSlots` (== `stack slot count`). Currently, the JIT stores `paramSlots` in `_currentThread->tempSlot`. 

Can the JIT store `paramElements` in another temp slot? It only needs to be set to a valid (> 0) value for the unresolved `invokedynamic` and `invokehandle` cases. For other cases, it can set to an invalid value.

nbhuiyan commented:
@babsingh, we can store the paramElements in another slot. Do you have any suggestions for what we can use given that `tempSlot` is already being used for num stack slots?

babsingh commented:
> Do you have any suggestions for what we can use given that tempSlot is already being used for num stack slots?

I am not sure. ++ @gacholio @tajila for a response.

babsingh commented:
> Do you have any suggestions for what we can use given that tempSlot is already being used for num stack slots?

From @tajila, 
https://github.com/eclipse/openj9/blob/c03a6f3242b966602ea812c268aab074db791bec/runtime/oti/j9nonbuilder.h#L4852-L4855
The above are volatile slots, which are used to pass arguments to runtime JIT helpers. We just need to make sure that they are unused for the `linkToStaticSpecial` code-path.

gacholio commented:
Yes, those slots should be available.

gacholio commented:
So is this ready to go or not?

babsingh commented:
> is this ready to go or not?

I have updated this PR to use `_currentThread->floatTemp1`. So, this PR can be reviewed.
Before it is merged, it will require the JIT changes, which @nazim is currently drafting.
I have changed this PR to WIP until the JIT changes are merged.

babsingh commented:
Updated the PR to remove the `appendix` object by shifting the arguments on the stack via `memmove`.

genie-openj9 commented:
Can one of the admins verify this patch?

babsingh commented:
@gacholio The dependent JIT and OMR changes have been merged. This PR is ready for a review.

gacholio commented:
jenkins test sanity winojdk292,xlinuxojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29

gacholio commented:
Windows javac is crashing.

babsingh commented:
Ref: https://github.com/eclipse-openj9/openj9/pull/12549#issuecomment-848813405. @nbhuiyan is working on a fix to resolve the build failures.

babsingh commented:
@nbhuiyan I have updated the PR to explicitly skip the stack adjustment when the JIT sets `floatTemp1` to a negative value. Even the previous change would have skipped the stack adjustment since `argDiff` would be negative for negative values of `floatTemp1`. But the previous change would skip implicitly.

I won't be able to test these changes without the dependent JIT changes. Also, I will on leave starting next Monday. If this PR does not get merged before my leave,
- you can either cherry-pick it in your PR and test it with the dependent JIT changes; or
- request support from the VM team (fyi @tajila @gacholio).

nbhuiyan commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -8362,8 +8362,28 @@ class INTERPRETER_CLASS
+			UDATA stackOffset = 1;
+			IDATA jitArgCount = (IDATA)_currentThread->floatTemp1;
+			/* Skip the below stack adjustment if JIT sets floatTemp1 to a negative value (-1). */
+			if (jitArgCount > 0) {
```
`jitArgCount` == 0 would mean the ROM method had 0 args, but this would still push the appendix object which may be NULL.

babsingh replied:
A `MemberName` is always expected here. So, `floatTemp1 (jitArgCount)` should never be 0 in `linkToStaticSpecial`.

Also, the `appendix object` will only exist for unresolved `invokedynamic` and `invokehandle` sequences. 

For skipping `stack adjustment`, the following condition will be true:
```
jitArgCount > 0 && methodArgCount > 0 && jitArgCount > methodArgCount
```

nbhuiyan commented:
@gacholio #12848 has been merged.

gacholio commented:
jenkins test sanity winojdk292,xlinuxojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29

gacholio commented:
```
20:10:16.671 0x1aa00    j9vm.224    *   ** ASSERTION FAILED ** at f:\users\jenkins\workspace\build_jdk16_x86-64_windows_ojdk292_personal\openj9\runtime\vm/BytecodeInterpreter.hpp:8379: ((argDiff == 1))
```

DanHeidinga commented:
@fengxue-IS Can you help with this?

fengxue-IS commented:
I will look into this
